dockstore-yaml-proposals/xena/.dockstore.yml

```diff
@@ -17,32 +18,60 @@ service:
      - stop.sh
 
   scripts:
-    # The next 5 scripts are always executed in the order listed here.
-    presprovision: # Before any provisioning
-    prestart: # After provisioning, before starting the service
-    start: stand_up.sh # Starts the service
-    poststart: # After the service has started
-    postprovision: load.sh # After the service has started. For multiple provisionings, e.g., the next day
+    # The scripts preprovision, prestart, start, poststart, and postprovision will be executed in order.
+    # The only script required to have a value is start.
+    
+    # Before any provisioning
+    preprovision: 
+    
+    # After provisioning, before starting the service
+    prestart: 
+     
+    # Starts the service
+    start: bash stand_up.sh 
+    
+    # After the service has started
+    poststart: 
+    
+    # After the service has started. For multiple provisionings, e.g., the user may decide to add more data a week later.
+    postprovision: bash load.sh 
 
-    port: port.sh # Returns the port the service is exposing.
-    healthcheck: # Returns
-    stop: stop.sh # Stops the service
+    # Returns the port the service is exposing. Can be used by a client to know what port to proxy
+    port: bash port.sh
+    
+    # Returns an exit code of 0 if service is running successfully.
+    healthcheck: 
+    
+    # Stops the service
+    stop: bash stop.sh 
 
-  # Parameters are passed as environment variables to all of the above scripts. As such, their names need to be valid environment variable names.
-  # The exception is parameters of type dataset and dataset[] -- those parameters are provisioned.
+  # Parameters are passed to the scripts as environment variables. They are service
+  # specific, i.e., the service creator decides which values to make parameterizable.
   parameters:
-    httpPort:
+    # The name of the parameter, can be anything that is a valid environment variable name
+    httpPort: 
+        # An optional default value
         default: "7222"
-        description: The host's HTTP port.
-    httpsPort:
-        default: "7223"
-        description: The host's HTTPS port
-    dataset:
-        type: dataset[]
+        # A description that tools can use for prompting the a user.
+        description: The host's HTTP port. The default is 7222.
+
+  # Data are files that are provisioned locally for the service. Each key is a dataset name,
+  # where the dataset can contain 1 to n files
+  data:
+    dataset_1:
+        targetDirectory: xena/files # Can override this at the individual file level.
```

Comment:
So from the presentation:
`xena/files` is a bit confusing, it doesn't make clear whether this intentionally matches "files" on line 42/70 below.

Reply:
It only matches by coincidence. Xena requires a directory named `xena/files` off of where Xena is running. Unfortunately I can't change that directory name (it's not configurable in Xena, AFAIK), and I think files is the right name for line 42/70, but I see how this confusing.

I'll try to clarify with comments.

Reply:
This is obviously just my preference, but since I'm more familiar with swagger.  I'd rather it be like:

```
parameters: 
  httpPort:
    default: "7222"
    type: string
    description: The host's HTTP port. The default is 7222.
  data:
    type: array
    items: 
      type: object
      properties:
        tsv:
          type: string
          description: Data for Xena in TSV format
        metadata:
          type: string
          description: The metadata, in JSON format, that corresponds to the data
```
It's much more verbose, but since the people filling in the JSON aren't really concerned with writing the parameters spec, I think it's alright.

I also don't see the need to make dataset some special type (that also still requires a definition).  The user and developer just needs to make it clear (via descriptions) what it's used for

The JSON would look like your original.

```
{
  "httpPort": "9001",
  "data": [
    {
      "tsv": "https://xena.treehouse.gi.ucsc.edu/download/TreehousePEDv9_Ribodeplete_clinical_metadata.2019-03-25.tsv",
      "metadata": "https://xena.treehouse.gi.ucsc.edu/download/TreehousePEDv9_Ribodeplete_clinical_metadata.2019-03-25.tsv.json"
    },
    {
      "tsv": "https://xena.treehouse.gi.ucsc.edu/download/TreehousePEDv9_Ribodeplete_unique_ensembl_expected_count.2019-03-25.tsv",
      "metadata": "https://xena.treehouse.gi.ucsc.edu/download/TreehousePEDv9_Ribodeplete_unique_ensembl_expected_count.2019-03-25.tsv.json"
    },
    {
      "tsv": "https://xena.treehouse.gi.ucsc.edu/download/gencode.v23.annotation.gene.probemap",
      "metadata": "https://xena.treehouse.gi.ucsc.edu/download/gencode.v23.annotation.gene.probemap.json"
    }
  ]
}
```

Reply:
First thought:

I initially had data as a child of parameters, but moved it out for this PR to be a sibling of parameters. The reason is that parameters are being passed as environment variables to the scripts, and I don't think it makes sense for data, especially with the nesting and arrays.

So I moved the data to be a sibling of parameters, rather than a child. That gives the advantage:

1. No longer need to declare types
2. Don't have to do special cases, e.g., some parameters are passed as environment variables and others aren't. Now all parameters are passed as environment variables, and data is treated separately.

Reply:
As the user (who is constructing the JSON), I don't think I care about implementation details like that.  But if you really did care, then it should be made more clear that parameters are environment variables only and data is not

Reply:
Good point, I'll call it `environmentVariables` or something like that instead of `parameters` to make it clear what it's being used for.